<a href="https://colab.research.google.com/github/Akanerd/Technical-Orbit/blob/main/Muhammad%20Hafizh%20Azzasafah_Jatayu%20-%20M05_S2_SQL.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Import Packages

In [ ]:
import sqlite3
import pandas as pd
import os
import shutil

# [A] Data Definition Language (DDL)

In [ ]:
# Setup
if os.path.exists("databaseku.db"):
    os.remove("databaseku.db")
else:
    print("The file does not exist")

## [A.1] Buat Koneksi ke Database

In [ ]:
connection = sqlite3.connect('databaseku.db')

## [A.2] Lihat Daftar Tabel

### Query SQL dan Fungsi Python untuk Melihat Daftar Tabel

In [ ]:
def lihat_daftar_tabel(con):
    cursorObj = con.cursor()
    cursorObj.execute('SELECT name from sqlite_master where type= "table"')
    print(cursorObj.fetchall())

### Contoh Database yang belum ada Tabel

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('databaseku.db')

# Panggil Fungsi untuk Lihat Daftar Tabel
lihat_daftar_tabel(connection)
connection.close()

[]


### Contoh Database yang sudah ada Tabel

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_penjualan.db')

# Panggil Fungsi untuk Lihat Daftar Tabel
lihat_daftar_tabel(connection)
connection.close()

[('orders',), ('sqlite_sequence',), ('agents',), ('customer',)]


## [A.3] CREATE
Membuat Tabel Baru

### Buat Tabel Baru

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('databaseku.db')

# Buat Query SQL
result = connection.cursor()
command = """
            CREATE TABLE siswa (
               id_siswa INTEGER PRIMARY KEY,
               nama TEXT NOT NULL,
               umur INTEGER NOT NULL
            );
          """
# Eksekusi Query SQL
result.execute(command)
connection.close()

### Lihat Daftar Tabel

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('databaseku.db')

# Panggil Fungsi untuk Lihat Daftar Tabel
lihat_daftar_tabel(connection)
connection.close()

[('siswa',)]


### Lihat Struktur Tabel

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('databaseku.db')

# Buat Query SQL
result  = connection.cursor()
command = "PRAGMA table_info('siswa');"

# Eksekusi Query SQL
result.execute(command)

# Tampilkan Hasil dalam Bentuk DataFrame
results      = result.fetchall();connection.close();
cols         = [column[0] for column in result.description]
results_data = pd.DataFrame.from_records(results, columns = cols)

results_data

cid      name     type  notnull dflt_value  pk
0    0  id_siswa  INTEGER        0       None   1
1    1      nama     TEXT        1       None   0
2    2      umur  INTEGER        1       None   0

## [A.4] ALTER
Mengubah struktur tabel

### Menambah Kolom pada Tabel

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('databaseku.db')

# Buat Query SQL
result = connection.cursor()
command = """
            ALTER TABLE siswa 
                ADD COLUMN kelas TEXT;
          """
# Eksekusi Query SQL
result.execute(command)
connection.close()


### Lihat Struktur Tabel

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('databaseku.db')

# Buat Query SQL
result  = connection.cursor()
command = "PRAGMA table_info('siswa');"

# Eksekusi Query SQL
result.execute(command)

# Tampilkan Hasil dalam Bentuk DataFrame
results      = result.fetchall();connection.close();
cols         = [column[0] for column in result.description]
results_data = pd.DataFrame.from_records(results, columns = cols)

results_data

cid      name     type  notnull dflt_value  pk
0    0  id_siswa  INTEGER        0       None   1
1    1      nama     TEXT        1       None   0
2    2      umur  INTEGER        1       None   0
3    3     kelas     TEXT        0       None   0

## [A.5] TRUNCATE (DELETE)
Menghapus semua data di tabel <br>
Catatan : Sqlite3 menggunakan DELETE bukan TRUNCATE untuk menghapus semua data di tabel

In [ ]:
# Setup
if os.path.exists("data_siswa.db"):
    os.remove("data_siswa.db")
else:
    print("The file does not exist")

shutil.copyfile("data_siswa_ori.db", "data_siswa.db")

'data_siswa.db'

### Lihat Isi Tabel

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_siswa.db')

# Buat Query SQL
result  = connection.cursor()
command = "SELECT * FROM siswa;"

# Eksekusi Query SQL
result.execute(command)

# Tampilkan Hasil dalam Bentuk DataFrame
results      = result.fetchall();connection.close();
cols         = [column[0] for column in result.description]
results_data = pd.DataFrame.from_records(results, columns = cols)

results_data

id_siswa   nama  umur    kelas
0         1   Budi    21  kelas A
1         2    Ani    20  kelas A
2         3   Tono    19  kelas B
3         4  Putra    22  kelas C
4         5  Putri    21  kelas B

### Menghapus Semua Data pada Tabel

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_siswa.db')

# Buat Query SQL
result = connection.cursor()
command = """
            DELETE FROM siswa;
          """
# Eksekusi Query SQL
result.execute(command)
connection.commit()
connection.close()

### Lihat Isi Tabel Kembali

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_siswa.db')

# Buat Query SQL
result  = connection.cursor()
command = "SELECT * FROM siswa;"

# Eksekusi Query SQL
result.execute(command)

# Tampilkan Hasil dalam Bentuk DataFrame
results      = result.fetchall();connection.close();
cols         = [column[0] for column in result.description]
results_data = pd.DataFrame.from_records(results, columns = cols)

results_data

Empty DataFrame
Columns: [id_siswa, nama, umur, kelas]
Index: []

### Lihat Daftar Tabel

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_siswa.db')

# Panggil Fungsi untuk Lihat Daftar Tabel
lihat_daftar_tabel(connection)
connection.close()

[('siswa',)]


## [A.6] DROP
Menghapus Tabel <br>

In [ ]:
# Setup
if os.path.exists("data_siswa.db"):
    os.remove("data_siswa.db")
else:
    print("The file does not exist")

shutil.copyfile("data_siswa_ori.db", "data_siswa.db")

'data_siswa.db'

### Lihat Daftar Tabel

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_siswa.db')

# Panggil Fungsi untuk Lihat Daftar Tabel
lihat_daftar_tabel(connection)
connection.close()

[('siswa',)]


### Menghapus Tabel

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_siswa.db')

# Buat Query SQL
result = connection.cursor()
command = """
            DROP TABLE siswa;
          """
# Eksekusi Query SQL
result.execute(command)
connection.commit()
connection.close()

### Lihat Daftar Tabel Kembali

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_siswa.db')

# Panggil Fungsi untuk Lihat Daftar Tabel
lihat_daftar_tabel(connection)
connection.close()

[]


# [B] Data Manipulation Language (DML)

In [ ]:
# Setup
if os.path.exists("data_penjualan.db"):
    os.remove("data_penjualan.db")
else:
    print("The file does not exist")

shutil.copyfile("data_penjualan_ori.db", "data_penjualan.db")

'data_penjualan.db'

## [B.1] SELECT

### Menampilkan Semua Data pada Tabel Agents

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_penjualan.db')

# Buat Query SQL
result  = connection.cursor()
command = "SELECT * FROM agents;"

# Eksekusi Query SQL
result.execute(command)

# Tampilkan Hasil dalam Bentuk DataFrame
results      = result.fetchall();connection.close();
cols         = [column[0] for column in result.description]
results_data = pd.DataFrame.from_records(results, columns = cols)

results_data

AGENT_CODE  AGENT_NAME WORKING_AREA  COMMISSION      PHONE_NO COUNTRY
0        A001    Subbarao    Bangalore        0.14  077-12346674        
1        A002      Mukesh       Mumbai        0.11  029-12358964        
2        A003       Alex        London        0.13  075-12458969        
3        A004        Ivan      Torento        0.15  008-22544166        
4        A005    Anderson      Brisban        0.13  045-21447739        
5        A006       McDen       London        0.15  078-22255588        
6        A007  Ramasundar    Bangalore        0.15  077-25814763        
7        A008      Alford     New York        0.12  044-25874365        
8        A009    Benjamin    Hampshair        0.11  008-22536178        
9        A010  Santakumar      Chennai        0.14  007-22388644        
10       A011  Ravi Kumar    Bangalore        0.15  077-45625874        
11       A012      Lucida     San Jose        0.12  044-52981425

### Menampilkan Semua Data pada Tabel Orders

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_penjualan.db')

# Buat Query SQL
result  = connection.cursor()
command = "SELECT * FROM orders;"

# Eksekusi Query SQL
result.execute(command)

# Tampilkan Hasil dalam Bentuk DataFrame
results      = result.fetchall();connection.close();
cols         = [column[0] for column in result.description]
results_data = pd.DataFrame.from_records(results, columns = cols)

results_data

ORD_NUM  ORD_AMOUNT  ADVANCE_AMOUNT ORD_DATE CUST_CODE AGENT_CODE  \
0    200100      1000.0           600.0             C00013       A003   
1    200101      3000.0          1000.0             C00001       A008   
2    200102      2000.0           300.0             C00012       A012   
3    200103      1500.0           700.0             C00021       A005   
4    200104      1500.0           500.0             C00006       A004   
5    200105      2500.0           500.0             C00025       A011   
6    200106      2500.0           700.0             C00005       A002   
7    200107      4500.0           900.0             C00007       A010   
8    200108      4000.0           600.0             C00008       A004   
9    200109      3500.0           800.0             C00011       A010   
10   200110      3000.0           500.0             C00019       A010   
11   200111      1000.0           300.0             C00020       A008   
12   200112      2000.0           400.0             C00016       A007   
13   200113      4000.0           600.0             C00022       A002   
14   200114      3500.0          2000.0             C00002       A008   
15   200116       500.0           100.0             C00010       A009   
16   200117       800.0           200.0             C00014       A001   
17   200118       500.0           100.0             C00023       A006   
18   200119      4000.0           700.0             C00007       A010   
19   200120       500.0           100.0             C00009       A002   
20   200121      1500.0           600.0             C00008       A004   
21   200122      2500.0           400.0             C00003       A004   
22   200123       500.0           100.0             C00022       A002   
23   200124       500.0           100.0             C00017       A007   
24   200125      2000.0           600.0             C00018       A005   
25   200126       500.0           100.0             C00022       A002   
26   200127      2500.0           400.0             C00015       A003   
27   200128      3500.0          1500.0             C00009       A002   
28   200129      2500.0           500.0             C00024       A006   
29   200130      2500.0           400.0             C00025       A011   
30   200131       900.0           150.0             C00012       A012   
31   200133      1200.0           400.0             C00009       A002   
32   200134      4200.0          1800.0             C00004       A005   
33   200135      2000.0           800.0             C00007       A010   

   ORD_DESCRIPTION  
0              SOD  
1              SOD  
2              SOD  
3              SOD  
4              SOD  
5              SOD  
6              SOD  
7              SOD  
8              SOD  
9              SOD  
10             SOD  
11             SOD  
12             SOD  
13             SOD  
14             SOD  
15             SOD  
16             SOD  
17             SOD  
18             SOD  
19             SOD  
20             SOD  
21             SOD  
22             SOD  
23             SOD  
24             SOD  
25             SOD  
26             SOD  
27             SOD  
28             SOD  
29             SOD  
30             SOD  
31             SOD  
32             SOD  
33             SOD

### Menampilkan Semua Data pada Tabel Customer

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_penjualan.db')

# Buat Query SQL
result  = connection.cursor()
command = "SELECT * FROM customer;"

# Eksekusi Query SQL
result.execute(command)

# Tampilkan Hasil dalam Bentuk DataFrame
results      = result.fetchall();connection.close();
cols         = [column[0] for column in result.description]
results_data = pd.DataFrame.from_records(results, columns = cols)

results_data

CUST_CODE    CUST_NAME  CUST_CITY WORKING_AREA CUST_COUNTRY  GRADE  \
0     C00001      Micheal   New York     New York          USA      2   
1     C00002         Bolt   New York     New York          USA      3   
2     C00003       Martin    Torento      Torento       Canada      2   
3     C00004      Winston    Brisban      Brisban    Australia      1   
4     C00005     Sasikant     Mumbai       Mumbai        India      1   
5     C00006      Shilton    Torento      Torento       Canada      1   
6     C00007   Ramanathan    Chennai      Chennai        India      1   
7     C00008     Karolina    Torento      Torento       Canada      1   
8     C00009       Ramesh     Mumbai       Mumbai        India      3   
9     C00010      Charles  Hampshair    Hampshair           UK      3   
10    C00011    Sundariya    Chennai      Chennai        India      3   
11    C00012       Steven   San Jose     San Jose          USA      1   
12    C00013       Holmes     London       London           UK      2   
13    C00014   Rangarappa  Bangalore    Bangalore        India      2   
14    C00015       Stuart     London       London           UK      1   
15    C00016   Venkatpati  Bangalore    Bangalore        India      2   
16    C00017     Srinivas  Bangalore    Bangalore        India      2   
17    C00018      Fleming    Brisban      Brisban    Australia      2   
18    C00019  Yearannaidu    Chennai      Chennai        India      1   
19    C00020       Albert   New York     New York          USA      3   
20    C00021        Jacks    Brisban      Brisban    Australia      1   
21    C00022      Avinash     Mumbai       Mumbai        India      2   
22    C00023         Karl     London       London           UK      0   
23    C00024         Cook     London       London           UK      2   
24    C00025    Ravindran  Bangalore    Bangalore        India      2   

    OPENING_AMT  RECEIVE_AMT  PAYMENT_AMT  OUTSTANDING_AMT      PHONE_NO  \
0        3000.0       5000.0       2000.0           6000.0       CCCCCCC   
1        5000.0       7000.0       9000.0           3000.0       DDNRDRH   
2        8000.0       7000.0       7000.0           8000.0       MJYURFD   
3        5000.0       8000.0       7000.0           6000.0       AAAAAAA   
4        7000.0      11000.0       7000.0          11000.0  147-25896312   
5       10000.0       7000.0       6000.0          11000.0       DDDDDDD   
6        7000.0      11000.0       9000.0           9000.0       GHRDWSD   
7        7000.0       7000.0       9000.0           5000.0       HJKORED   
8        8000.0       7000.0       3000.0          12000.0      Phone No   
9        6000.0       4000.0       5000.0           5000.0       MMMMMMM   
10       7000.0      11000.0       7000.0          11000.0       PPHGRTS   
11       5000.0       7000.0       9000.0           3000.0       KRFYGJK   
12       6000.0       5000.0       7000.0           4000.0       BBBBBBB   
13       8000.0      11000.0       7000.0          12000.0       AAAATGF   
14       6000.0       8000.0       3000.0          11000.0       GFSGERS   
15       8000.0      11000.0       7000.0          12000.0       JRTVFDD   
16       8000.0       4000.0       3000.0           9000.0       AAAAAAB   
17       7000.0       7000.0       9000.0           5000.0       NHBGVFC   
18       8000.0       7000.0       7000.0           8000.0       ZZZZBFV   
19       5000.0       7000.0       6000.0           6000.0       BBBBSBB   
20       7000.0       7000.0       7000.0           7000.0       WERTGDF   
21       7000.0      11000.0       9000.0           9000.0  113-12345678   
22       4000.0       6000.0       7000.0           3000.0       AAAABAA   
23       4000.0       9000.0       7000.0           6000.0       FSDDSDF   
24       5000.0       7000.0       4000.0           8000.0       AVAVAVA   

   AGENT_CODE  
0        A008  
1        A008  
2        A004  
3        A005  
4        A002  
5        A004  
6        A010  

## [B.1] SELECT (Contoh)
Contoh : menampilkan sebagian/beberapa kolom pada  tabel Customer.

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_penjualan.db')

# Buat Query SQL
result  = connection.cursor()
command = "SELECT CUST_NAME,CUST_CITY AS 'Kota Asal' FROM customer;"

# Eksekusi Query SQL
result.execute(command)

# Tampilkan Hasil dalam Bentuk DataFrame
results      = result.fetchall();connection.close();
cols         = [column[0] for column in result.description]
results_data = pd.DataFrame.from_records(results, columns = cols)

results_data

CUST_NAME  Kota Asal
0       Micheal   New York
1          Bolt   New York
2        Martin    Torento
3       Winston    Brisban
4      Sasikant     Mumbai
5       Shilton    Torento
6    Ramanathan    Chennai
7      Karolina    Torento
8        Ramesh     Mumbai
9       Charles  Hampshair
10    Sundariya    Chennai
11       Steven   San Jose
12       Holmes     London
13   Rangarappa  Bangalore
14       Stuart     London
15   Venkatpati  Bangalore
16     Srinivas  Bangalore
17      Fleming    Brisban
18  Yearannaidu    Chennai
19       Albert   New York
20        Jacks    Brisban
21      Avinash     Mumbai
22         Karl     London
23         Cook     London
24    Ravindran  Bangalore

## [B.2] INSERT

### Menambahkan Data pada Tabel Agents

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_penjualan.db')

# Buat Query SQL
result  = connection.cursor()
command = """
            INSERT INTO agents (AGENT_CODE,AGENT_NAME,WORKING_AREA,COMMISSION,PHONE_NO,COUNTRY) 
                VALUES ('A013', 'Achmad Malik', 'Indonesia', '0.27', '021-14045', '');
          """

# Eksekusi Query SQL
result.execute(command)
connection.commit()
connection.close()

### Menampilkan Semua Data pada Tabel Agents

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_penjualan.db')

# Buat Query SQL
result  = connection.cursor()
command = "SELECT * FROM agents;"

# Eksekusi Query SQL
result.execute(command)

# Tampilkan Hasil dalam Bentuk DataFrame
results      = result.fetchall();connection.close();
cols         = [column[0] for column in result.description]
results_data = pd.DataFrame.from_records(results, columns = cols)

results_data

AGENT_CODE    AGENT_NAME WORKING_AREA  COMMISSION      PHONE_NO COUNTRY
0        A001      Subbarao    Bangalore        0.14  077-12346674        
1        A002        Mukesh       Mumbai        0.11  029-12358964        
2        A003         Alex        London        0.13  075-12458969        
3        A004          Ivan      Torento        0.15  008-22544166        
4        A005      Anderson      Brisban        0.13  045-21447739        
5        A006         McDen       London        0.15  078-22255588        
6        A007    Ramasundar    Bangalore        0.15  077-25814763        
7        A008        Alford     New York        0.12  044-25874365        
8        A009      Benjamin    Hampshair        0.11  008-22536178        
9        A010    Santakumar      Chennai        0.14  007-22388644        
10       A011    Ravi Kumar    Bangalore        0.15  077-45625874        
11       A012        Lucida     San Jose        0.12  044-52981425        
12       A013  Achmad Malik    Indonesia        0.27     021-14045

## [B.3] UPDATE

### Menampilkan Semua Data pada Tabel Customer (Sebelum Update)

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_penjualan.db')

# Buat Query SQL
result  = connection.cursor()
command = "SELECT CUST_NAME,GRADE FROM customer;"

# Eksekusi Query SQL
result.execute(command)

# Tampilkan Hasil dalam Bentuk DataFrame
results      = result.fetchall();connection.close();
cols         = [column[0] for column in result.description]
results_data = pd.DataFrame.from_records(results, columns = cols)

results_data

CUST_NAME  GRADE
0       Micheal      2
1          Bolt      3
2        Martin      2
3       Winston      1
4      Sasikant      1
5       Shilton      1
6    Ramanathan      1
7      Karolina      1
8        Ramesh      3
9       Charles      3
10    Sundariya      3
11       Steven      1
12       Holmes      2
13   Rangarappa      2
14       Stuart      1
15   Venkatpati      2
16     Srinivas      2
17      Fleming      2
18  Yearannaidu      1
19       Albert      3
20        Jacks      1
21      Avinash      2
22         Karl      0
23         Cook      2
24    Ravindran      2

### Melakukan Update pada Tabel Customer

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_penjualan.db')

# Buat Query SQL
result  = connection.cursor()
command = "UPDATE customer SET GRADE = 1 WHERE CUST_NAME = 'Micheal'"

# Eksekusi Query SQL
result.execute(command)
connection.commit()
connection.close()

### Menampilkan Semua Data pada Tabel Customer (Setelah Update)

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_penjualan.db')

# Buat Query SQL
result  = connection.cursor()
command = "SELECT CUST_NAME,GRADE FROM customer;"

# Eksekusi Query SQL
result.execute(command)

# Tampilkan Hasil dalam Bentuk DataFrame
results      = result.fetchall();connection.close();
cols         = [column[0] for column in result.description]
results_data = pd.DataFrame.from_records(results, columns = cols)

results_data

CUST_NAME  GRADE
0       Micheal      1
1          Bolt      3
2        Martin      2
3       Winston      1
4      Sasikant      1
5       Shilton      1
6    Ramanathan      1
7      Karolina      1
8        Ramesh      3
9       Charles      3
10    Sundariya      3
11       Steven      1
12       Holmes      2
13   Rangarappa      2
14       Stuart      1
15   Venkatpati      2
16     Srinivas      2
17      Fleming      2
18  Yearannaidu      1
19       Albert      3
20        Jacks      1
21      Avinash      2
22         Karl      0
23         Cook      2
24    Ravindran      2

## [B.4] DELETE

### Menampilkan Semua Data pada Tabel Customer (Sebelum Delete)

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_penjualan.db')

# Buat Query SQL
result  = connection.cursor()
command = "SELECT * FROM customer;"

# Eksekusi Query SQL
result.execute(command)

# Tampilkan Hasil dalam Bentuk DataFrame
results      = result.fetchall();connection.close();
cols         = [column[0] for column in result.description]
results_data = pd.DataFrame.from_records(results, columns = cols)

results_data

CUST_CODE    CUST_NAME  CUST_CITY WORKING_AREA CUST_COUNTRY  GRADE  \
0     C00001      Micheal   New York     New York          USA      1   
1     C00002         Bolt   New York     New York          USA      3   
2     C00003       Martin    Torento      Torento       Canada      2   
3     C00004      Winston    Brisban      Brisban    Australia      1   
4     C00005     Sasikant     Mumbai       Mumbai        India      1   
5     C00006      Shilton    Torento      Torento       Canada      1   
6     C00007   Ramanathan    Chennai      Chennai        India      1   
7     C00008     Karolina    Torento      Torento       Canada      1   
8     C00009       Ramesh     Mumbai       Mumbai        India      3   
9     C00010      Charles  Hampshair    Hampshair           UK      3   
10    C00011    Sundariya    Chennai      Chennai        India      3   
11    C00012       Steven   San Jose     San Jose          USA      1   
12    C00013       Holmes     London       London           UK      2   
13    C00014   Rangarappa  Bangalore    Bangalore        India      2   
14    C00015       Stuart     London       London           UK      1   
15    C00016   Venkatpati  Bangalore    Bangalore        India      2   
16    C00017     Srinivas  Bangalore    Bangalore        India      2   
17    C00018      Fleming    Brisban      Brisban    Australia      2   
18    C00019  Yearannaidu    Chennai      Chennai        India      1   
19    C00020       Albert   New York     New York          USA      3   
20    C00021        Jacks    Brisban      Brisban    Australia      1   
21    C00022      Avinash     Mumbai       Mumbai        India      2   
22    C00023         Karl     London       London           UK      0   
23    C00024         Cook     London       London           UK      2   
24    C00025    Ravindran  Bangalore    Bangalore        India      2   

    OPENING_AMT  RECEIVE_AMT  PAYMENT_AMT  OUTSTANDING_AMT      PHONE_NO  \
0        3000.0       5000.0       2000.0           6000.0       CCCCCCC   
1        5000.0       7000.0       9000.0           3000.0       DDNRDRH   
2        8000.0       7000.0       7000.0           8000.0       MJYURFD   
3        5000.0       8000.0       7000.0           6000.0       AAAAAAA   
4        7000.0      11000.0       7000.0          11000.0  147-25896312   
5       10000.0       7000.0       6000.0          11000.0       DDDDDDD   
6        7000.0      11000.0       9000.0           9000.0       GHRDWSD   
7        7000.0       7000.0       9000.0           5000.0       HJKORED   
8        8000.0       7000.0       3000.0          12000.0      Phone No   
9        6000.0       4000.0       5000.0           5000.0       MMMMMMM   
10       7000.0      11000.0       7000.0          11000.0       PPHGRTS   
11       5000.0       7000.0       9000.0           3000.0       KRFYGJK   
12       6000.0       5000.0       7000.0           4000.0       BBBBBBB   
13       8000.0      11000.0       7000.0          12000.0       AAAATGF   
14       6000.0       8000.0       3000.0          11000.0       GFSGERS   
15       8000.0      11000.0       7000.0          12000.0       JRTVFDD   
16       8000.0       4000.0       3000.0           9000.0       AAAAAAB   
17       7000.0       7000.0       9000.0           5000.0       NHBGVFC   
18       8000.0       7000.0       7000.0           8000.0       ZZZZBFV   
19       5000.0       7000.0       6000.0           6000.0       BBBBSBB   
20       7000.0       7000.0       7000.0           7000.0       WERTGDF   
21       7000.0      11000.0       9000.0           9000.0  113-12345678   
22       4000.0       6000.0       7000.0           3000.0       AAAABAA   
23       4000.0       9000.0       7000.0           6000.0       FSDDSDF   
24       5000.0       7000.0       4000.0           8000.0       AVAVAVA   

   AGENT_CODE  
0        A008  
1        A008  
2        A004  
3        A005  
4        A002  
5        A004  
6        A010  

### Menghapus Data dengan Kriteria Tertentu pada Tabel Customer

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_penjualan.db')

# Buat Query SQL
result  = connection.cursor()
command = "DELETE FROM customer WHERE CUST_NAME = 'Micheal';"

# Eksekusi Query SQL
result.execute(command)
connection.commit()
connection.close()

### Menampilkan Semua Data pada Tabel Customer (Setelah Delete)

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_penjualan.db')

# Buat Query SQL
result  = connection.cursor()
command = "SELECT * FROM customer;"

# Eksekusi Query SQL
result.execute(command)

# Tampilkan Hasil dalam Bentuk DataFrame
results      = result.fetchall();connection.close();
cols         = [column[0] for column in result.description]
results_data = pd.DataFrame.from_records(results, columns = cols)

results_data

CUST_CODE    CUST_NAME  CUST_CITY WORKING_AREA CUST_COUNTRY  GRADE  \
0     C00002         Bolt   New York     New York          USA      3   
1     C00003       Martin    Torento      Torento       Canada      2   
2     C00004      Winston    Brisban      Brisban    Australia      1   
3     C00005     Sasikant     Mumbai       Mumbai        India      1   
4     C00006      Shilton    Torento      Torento       Canada      1   
5     C00007   Ramanathan    Chennai      Chennai        India      1   
6     C00008     Karolina    Torento      Torento       Canada      1   
7     C00009       Ramesh     Mumbai       Mumbai        India      3   
8     C00010      Charles  Hampshair    Hampshair           UK      3   
9     C00011    Sundariya    Chennai      Chennai        India      3   
10    C00012       Steven   San Jose     San Jose          USA      1   
11    C00013       Holmes     London       London           UK      2   
12    C00014   Rangarappa  Bangalore    Bangalore        India      2   
13    C00015       Stuart     London       London           UK      1   
14    C00016   Venkatpati  Bangalore    Bangalore        India      2   
15    C00017     Srinivas  Bangalore    Bangalore        India      2   
16    C00018      Fleming    Brisban      Brisban    Australia      2   
17    C00019  Yearannaidu    Chennai      Chennai        India      1   
18    C00020       Albert   New York     New York          USA      3   
19    C00021        Jacks    Brisban      Brisban    Australia      1   
20    C00022      Avinash     Mumbai       Mumbai        India      2   
21    C00023         Karl     London       London           UK      0   
22    C00024         Cook     London       London           UK      2   
23    C00025    Ravindran  Bangalore    Bangalore        India      2   

    OPENING_AMT  RECEIVE_AMT  PAYMENT_AMT  OUTSTANDING_AMT      PHONE_NO  \
0        5000.0       7000.0       9000.0           3000.0       DDNRDRH   
1        8000.0       7000.0       7000.0           8000.0       MJYURFD   
2        5000.0       8000.0       7000.0           6000.0       AAAAAAA   
3        7000.0      11000.0       7000.0          11000.0  147-25896312   
4       10000.0       7000.0       6000.0          11000.0       DDDDDDD   
5        7000.0      11000.0       9000.0           9000.0       GHRDWSD   
6        7000.0       7000.0       9000.0           5000.0       HJKORED   
7        8000.0       7000.0       3000.0          12000.0      Phone No   
8        6000.0       4000.0       5000.0           5000.0       MMMMMMM   
9        7000.0      11000.0       7000.0          11000.0       PPHGRTS   
10       5000.0       7000.0       9000.0           3000.0       KRFYGJK   
11       6000.0       5000.0       7000.0           4000.0       BBBBBBB   
12       8000.0      11000.0       7000.0          12000.0       AAAATGF   
13       6000.0       8000.0       3000.0          11000.0       GFSGERS   
14       8000.0      11000.0       7000.0          12000.0       JRTVFDD   
15       8000.0       4000.0       3000.0           9000.0       AAAAAAB   
16       7000.0       7000.0       9000.0           5000.0       NHBGVFC   
17       8000.0       7000.0       7000.0           8000.0       ZZZZBFV   
18       5000.0       7000.0       6000.0           6000.0       BBBBSBB   
19       7000.0       7000.0       7000.0           7000.0       WERTGDF   
20       7000.0      11000.0       9000.0           9000.0  113-12345678   
21       4000.0       6000.0       7000.0           3000.0       AAAABAA   
22       4000.0       9000.0       7000.0           6000.0       FSDDSDF   
23       5000.0       7000.0       4000.0           8000.0       AVAVAVA   

   AGENT_CODE  
0        A008  
1        A004  
2        A005  
3        A002  
4        A004  
5        A010  
6        A004  
7        A002  
8        A009  
9        A010  
10       A012  
11       A003  
12       A001  
13       A003  
14       A007  
15       A007  
16  

# [C] Filtering dan Subsetting

In [ ]:
# Setup
if os.path.exists("data_penjualan.db"):
    os.remove("data_penjualan.db")
else:
    print("The file does not exist")

shutil.copyfile("data_penjualan_ori.db", "data_penjualan.db")

'data_penjualan.db'

## [C.1] Operator Perbandingan

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_penjualan.db')

# Membuat Query SQL
result  = connection.cursor()
command = "SELECT * FROM agents WHERE AGENT_CODE = 'A002';"

# Eksekusi Query SQL
result.execute(command)

# Menampilkan Hasil Dalam Bentuk DataFrame
results      = result.fetchall();connection.close();
cols         = [column[0] for column in result.description]
results_data = pd.DataFrame.from_records(results, columns = cols)

results_data

AGENT_CODE AGENT_NAME WORKING_AREA  COMMISSION      PHONE_NO COUNTRY
0       A002     Mukesh       Mumbai        0.11  029-12358964

## [C.2] Operator Boolean

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_penjualan.db')

# Membuat Query SQL
result  = connection.cursor()
command = "SELECT * FROM orders WHERE ORD_AMOUNT >= 500 AND ADVANCE_AMOUNT >= 300;"

# Eksekusi Query SQL
result.execute(command)

# Menampilkan Hasil Dalam Bentuk DataFrame
results      = result.fetchall();connection.close();
cols         = [column[0] for column in result.description]
results_data = pd.DataFrame.from_records(results, columns = cols)

results_data

ORD_NUM  ORD_AMOUNT  ADVANCE_AMOUNT ORD_DATE CUST_CODE AGENT_CODE  \
0    200100      1000.0           600.0             C00013       A003   
1    200101      3000.0          1000.0             C00001       A008   
2    200102      2000.0           300.0             C00012       A012   
3    200103      1500.0           700.0             C00021       A005   
4    200104      1500.0           500.0             C00006       A004   
5    200105      2500.0           500.0             C00025       A011   
6    200106      2500.0           700.0             C00005       A002   
7    200107      4500.0           900.0             C00007       A010   
8    200108      4000.0           600.0             C00008       A004   
9    200109      3500.0           800.0             C00011       A010   
10   200110      3000.0           500.0             C00019       A010   
11   200111      1000.0           300.0             C00020       A008   
12   200112      2000.0           400.0             C00016       A007   
13   200113      4000.0           600.0             C00022       A002   
14   200114      3500.0          2000.0             C00002       A008   
15   200119      4000.0           700.0             C00007       A010   
16   200121      1500.0           600.0             C00008       A004   
17   200122      2500.0           400.0             C00003       A004   
18   200125      2000.0           600.0             C00018       A005   
19   200127      2500.0           400.0             C00015       A003   
20   200128      3500.0          1500.0             C00009       A002   
21   200129      2500.0           500.0             C00024       A006   
22   200130      2500.0           400.0             C00025       A011   
23   200133      1200.0           400.0             C00009       A002   
24   200134      4200.0          1800.0             C00004       A005   
25   200135      2000.0           800.0             C00007       A010   

   ORD_DESCRIPTION  
0              SOD  
1              SOD  
2              SOD  
3              SOD  
4              SOD  
5              SOD  
6              SOD  
7              SOD  
8              SOD  
9              SOD  
10             SOD  
11             SOD  
12             SOD  
13             SOD  
14             SOD  
15             SOD  
16             SOD  
17             SOD  
18             SOD  
19             SOD  
20             SOD  
21             SOD  
22             SOD  
23             SOD  
24             SOD  
25             SOD

## [C.3] Operator LIKE

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_penjualan.db')

# Membuat Query SQL
result  = connection.cursor()
command =  "SELECT * FROM customer WHERE CUST_NAME LIKE 'S%';" # Nama dengan Awalan 'S'

# Eksekusi Query SQL
result.execute(command)

# Menampilkan Hasil Dalam Bentuk DataFrame
results      = result.fetchall();connection.close();
cols         = [column[0] for column in result.description]
results_data = pd.DataFrame.from_records(results, columns = cols)

results_data

CUST_CODE  CUST_NAME  CUST_CITY WORKING_AREA CUST_COUNTRY  GRADE  \
0    C00005   Sasikant     Mumbai       Mumbai        India      1   
1    C00006    Shilton    Torento      Torento       Canada      1   
2    C00011  Sundariya    Chennai      Chennai        India      3   
3    C00012     Steven   San Jose     San Jose          USA      1   
4    C00015     Stuart     London       London           UK      1   
5    C00017   Srinivas  Bangalore    Bangalore        India      2   

   OPENING_AMT  RECEIVE_AMT  PAYMENT_AMT  OUTSTANDING_AMT      PHONE_NO  \
0       7000.0      11000.0       7000.0          11000.0  147-25896312   
1      10000.0       7000.0       6000.0          11000.0       DDDDDDD   
2       7000.0      11000.0       7000.0          11000.0       PPHGRTS   
3       5000.0       7000.0       9000.0           3000.0       KRFYGJK   
4       6000.0       8000.0       3000.0          11000.0       GFSGERS   
5       8000.0       4000.0       3000.0           9000.0       AAAAAAB   

  AGENT_CODE  
0       A002  
1       A004  
2       A010  
3       A012  
4       A003  
5       A007

## [C.4] Operator IN dan NOT IN

### Operator IN

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_penjualan.db')

# Membuat Query SQL
result  = connection.cursor()
command = "SELECT * FROM customer WHERE CUST_COUNTRY IN ('USA','Canada','India');"

# Eksekusi Query SQL
result.execute(command)

# Menampilkan Hasil Dalam Bentuk DataFrame
results      = result.fetchall();connection.close();
cols         = [column[0] for column in result.description]
results_data = pd.DataFrame.from_records(results, columns = cols)

results_data

CUST_CODE    CUST_NAME  CUST_CITY WORKING_AREA CUST_COUNTRY  GRADE  \
0     C00001      Micheal   New York     New York          USA      2   
1     C00002         Bolt   New York     New York          USA      3   
2     C00003       Martin    Torento      Torento       Canada      2   
3     C00005     Sasikant     Mumbai       Mumbai        India      1   
4     C00006      Shilton    Torento      Torento       Canada      1   
5     C00007   Ramanathan    Chennai      Chennai        India      1   
6     C00008     Karolina    Torento      Torento       Canada      1   
7     C00009       Ramesh     Mumbai       Mumbai        India      3   
8     C00011    Sundariya    Chennai      Chennai        India      3   
9     C00012       Steven   San Jose     San Jose          USA      1   
10    C00014   Rangarappa  Bangalore    Bangalore        India      2   
11    C00016   Venkatpati  Bangalore    Bangalore        India      2   
12    C00017     Srinivas  Bangalore    Bangalore        India      2   
13    C00019  Yearannaidu    Chennai      Chennai        India      1   
14    C00020       Albert   New York     New York          USA      3   
15    C00022      Avinash     Mumbai       Mumbai        India      2   
16    C00025    Ravindran  Bangalore    Bangalore        India      2   

    OPENING_AMT  RECEIVE_AMT  PAYMENT_AMT  OUTSTANDING_AMT      PHONE_NO  \
0        3000.0       5000.0       2000.0           6000.0       CCCCCCC   
1        5000.0       7000.0       9000.0           3000.0       DDNRDRH   
2        8000.0       7000.0       7000.0           8000.0       MJYURFD   
3        7000.0      11000.0       7000.0          11000.0  147-25896312   
4       10000.0       7000.0       6000.0          11000.0       DDDDDDD   
5        7000.0      11000.0       9000.0           9000.0       GHRDWSD   
6        7000.0       7000.0       9000.0           5000.0       HJKORED   
7        8000.0       7000.0       3000.0          12000.0      Phone No   
8        7000.0      11000.0       7000.0          11000.0       PPHGRTS   
9        5000.0       7000.0       9000.0           3000.0       KRFYGJK   
10       8000.0      11000.0       7000.0          12000.0       AAAATGF   
11       8000.0      11000.0       7000.0          12000.0       JRTVFDD   
12       8000.0       4000.0       3000.0           9000.0       AAAAAAB   
13       8000.0       7000.0       7000.0           8000.0       ZZZZBFV   
14       5000.0       7000.0       6000.0           6000.0       BBBBSBB   
15       7000.0      11000.0       9000.0           9000.0  113-12345678   
16       5000.0       7000.0       4000.0           8000.0       AVAVAVA   

   AGENT_CODE  
0        A008  
1        A008  
2        A004  
3        A002  
4        A004  
5        A010  
6        A004  
7        A002  
8        A010  
9        A012  
10       A001  
11       A007  
12       A007  
13       A010  
14       A008  
15       A002  
16       A011

### Operator NOT IN

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_penjualan.db')

# Membuat Query SQL
result  = connection.cursor()
command = "SELECT * FROM customer WHERE CUST_COUNTRY NOT IN ('USA','Canada','India');"

# Eksekusi Query SQL
result.execute(command)

# Menampilkan Hasil Dalam Bentuk DataFrame
results      = result.fetchall();connection.close();
cols         = [column[0] for column in result.description]
results_data = pd.DataFrame.from_records(results, columns = cols)

results_data

CUST_CODE CUST_NAME  CUST_CITY WORKING_AREA CUST_COUNTRY  GRADE  \
0    C00004   Winston    Brisban      Brisban    Australia      1   
1    C00010   Charles  Hampshair    Hampshair           UK      3   
2    C00013    Holmes     London       London           UK      2   
3    C00015    Stuart     London       London           UK      1   
4    C00018   Fleming    Brisban      Brisban    Australia      2   
5    C00021     Jacks    Brisban      Brisban    Australia      1   
6    C00023      Karl     London       London           UK      0   
7    C00024      Cook     London       London           UK      2   

   OPENING_AMT  RECEIVE_AMT  PAYMENT_AMT  OUTSTANDING_AMT PHONE_NO AGENT_CODE  
0       5000.0       8000.0       7000.0           6000.0  AAAAAAA       A005  
1       6000.0       4000.0       5000.0           5000.0  MMMMMMM       A009  
2       6000.0       5000.0       7000.0           4000.0  BBBBBBB       A003  
3       6000.0       8000.0       3000.0          11000.0  GFSGERS       A003  
4       7000.0       7000.0       9000.0           5000.0  NHBGVFC       A005  
5       7000.0       7000.0       7000.0           7000.0  WERTGDF       A005  
6       4000.0       6000.0       7000.0           3000.0  AAAABAA       A006  
7       4000.0       9000.0       7000.0           6000.0  FSDDSDF       A006

## [C.5] Operator BETWEEN

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_penjualan.db')

# Membuat Query SQL
result  = connection.cursor()
command = "SELECT * FROM orders WHERE ADVANCE_AMOUNT BETWEEN '500' AND '800';"

# Eksekusi Query SQL
result.execute(command)

# Menampilkan Hasil Dalam Bentuk DataFrame
results      = result.fetchall();connection.close();
cols         = [column[0] for column in result.description]
results_data = pd.DataFrame.from_records(results, columns = cols)

results_data

ORD_NUM  ORD_AMOUNT  ADVANCE_AMOUNT ORD_DATE CUST_CODE AGENT_CODE  \
0    200100      1000.0           600.0             C00013       A003   
1    200103      1500.0           700.0             C00021       A005   
2    200104      1500.0           500.0             C00006       A004   
3    200105      2500.0           500.0             C00025       A011   
4    200106      2500.0           700.0             C00005       A002   
5    200108      4000.0           600.0             C00008       A004   
6    200109      3500.0           800.0             C00011       A010   
7    200110      3000.0           500.0             C00019       A010   
8    200113      4000.0           600.0             C00022       A002   
9    200119      4000.0           700.0             C00007       A010   
10   200121      1500.0           600.0             C00008       A004   
11   200125      2000.0           600.0             C00018       A005   
12   200129      2500.0           500.0             C00024       A006   
13   200135      2000.0           800.0             C00007       A010   

   ORD_DESCRIPTION  
0              SOD  
1              SOD  
2              SOD  
3              SOD  
4              SOD  
5              SOD  
6              SOD  
7              SOD  
8              SOD  
9              SOD  
10             SOD  
11             SOD  
12             SOD  
13             SOD

# [D] Sorting dan Grouping

In [ ]:
# Setup
if os.path.exists("data_penjualan.db"):
    os.remove("data_penjualan.db")
else:
    print("The file does not exist")

shutil.copyfile("data_penjualan_ori.db", "data_penjualan.db")

'data_penjualan.db'

##  [D.1] Mengurutkan Data (Sorting)

### Data pada Tabel Customer Sebelum Diurutkan

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_penjualan.db')

# Membuat Query SQL
result  = connection.cursor()
command = "SELECT * FROM customer;"

# Eksekusi Query SQL
result.execute(command)

# Menampilkan Hasil Dalam Bentuk DataFrame
results      = result.fetchall();connection.close();
cols         = [column[0] for column in result.description]
results_data = pd.DataFrame.from_records(results, columns = cols)

results_data

CUST_CODE    CUST_NAME  CUST_CITY WORKING_AREA CUST_COUNTRY  GRADE  \
0     C00001      Micheal   New York     New York          USA      2   
1     C00002         Bolt   New York     New York          USA      3   
2     C00003       Martin    Torento      Torento       Canada      2   
3     C00004      Winston    Brisban      Brisban    Australia      1   
4     C00005     Sasikant     Mumbai       Mumbai        India      1   
5     C00006      Shilton    Torento      Torento       Canada      1   
6     C00007   Ramanathan    Chennai      Chennai        India      1   
7     C00008     Karolina    Torento      Torento       Canada      1   
8     C00009       Ramesh     Mumbai       Mumbai        India      3   
9     C00010      Charles  Hampshair    Hampshair           UK      3   
10    C00011    Sundariya    Chennai      Chennai        India      3   
11    C00012       Steven   San Jose     San Jose          USA      1   
12    C00013       Holmes     London       London           UK      2   
13    C00014   Rangarappa  Bangalore    Bangalore        India      2   
14    C00015       Stuart     London       London           UK      1   
15    C00016   Venkatpati  Bangalore    Bangalore        India      2   
16    C00017     Srinivas  Bangalore    Bangalore        India      2   
17    C00018      Fleming    Brisban      Brisban    Australia      2   
18    C00019  Yearannaidu    Chennai      Chennai        India      1   
19    C00020       Albert   New York     New York          USA      3   
20    C00021        Jacks    Brisban      Brisban    Australia      1   
21    C00022      Avinash     Mumbai       Mumbai        India      2   
22    C00023         Karl     London       London           UK      0   
23    C00024         Cook     London       London           UK      2   
24    C00025    Ravindran  Bangalore    Bangalore        India      2   

    OPENING_AMT  RECEIVE_AMT  PAYMENT_AMT  OUTSTANDING_AMT      PHONE_NO  \
0        3000.0       5000.0       2000.0           6000.0       CCCCCCC   
1        5000.0       7000.0       9000.0           3000.0       DDNRDRH   
2        8000.0       7000.0       7000.0           8000.0       MJYURFD   
3        5000.0       8000.0       7000.0           6000.0       AAAAAAA   
4        7000.0      11000.0       7000.0          11000.0  147-25896312   
5       10000.0       7000.0       6000.0          11000.0       DDDDDDD   
6        7000.0      11000.0       9000.0           9000.0       GHRDWSD   
7        7000.0       7000.0       9000.0           5000.0       HJKORED   
8        8000.0       7000.0       3000.0          12000.0      Phone No   
9        6000.0       4000.0       5000.0           5000.0       MMMMMMM   
10       7000.0      11000.0       7000.0          11000.0       PPHGRTS   
11       5000.0       7000.0       9000.0           3000.0       KRFYGJK   
12       6000.0       5000.0       7000.0           4000.0       BBBBBBB   
13       8000.0      11000.0       7000.0          12000.0       AAAATGF   
14       6000.0       8000.0       3000.0          11000.0       GFSGERS   
15       8000.0      11000.0       7000.0          12000.0       JRTVFDD   
16       8000.0       4000.0       3000.0           9000.0       AAAAAAB   
17       7000.0       7000.0       9000.0           5000.0       NHBGVFC   
18       8000.0       7000.0       7000.0           8000.0       ZZZZBFV   
19       5000.0       7000.0       6000.0           6000.0       BBBBSBB   
20       7000.0       7000.0       7000.0           7000.0       WERTGDF   
21       7000.0      11000.0       9000.0           9000.0  113-12345678   
22       4000.0       6000.0       7000.0           3000.0       AAAABAA   
23       4000.0       9000.0       7000.0           6000.0       FSDDSDF   
24       5000.0       7000.0       4000.0           8000.0       AVAVAVA   

   AGENT_CODE  
0        A008  
1        A008  
2        A004  
3        A005  
4        A002  
5        A004  
6        A010  

### Data pada Tabel Customer Setelah Diurutkan

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_penjualan.db')

# Membuat Query SQL
result  = connection.cursor()
command = "SELECT * FROM customer ORDER BY CUST_NAME ASC;"

# Eksekusi Query SQL
result.execute(command)

# Menampilkan Hasil Dalam Bentuk DataFrame
results      = result.fetchall();connection.close();
cols         = [column[0] for column in result.description]
results_data = pd.DataFrame.from_records(results, columns = cols)

results_data

CUST_CODE    CUST_NAME  CUST_CITY WORKING_AREA CUST_COUNTRY  GRADE  \
0     C00020       Albert   New York     New York          USA      3   
1     C00022      Avinash     Mumbai       Mumbai        India      2   
2     C00002         Bolt   New York     New York          USA      3   
3     C00010      Charles  Hampshair    Hampshair           UK      3   
4     C00024         Cook     London       London           UK      2   
5     C00018      Fleming    Brisban      Brisban    Australia      2   
6     C00013       Holmes     London       London           UK      2   
7     C00021        Jacks    Brisban      Brisban    Australia      1   
8     C00023         Karl     London       London           UK      0   
9     C00008     Karolina    Torento      Torento       Canada      1   
10    C00003       Martin    Torento      Torento       Canada      2   
11    C00001      Micheal   New York     New York          USA      2   
12    C00007   Ramanathan    Chennai      Chennai        India      1   
13    C00009       Ramesh     Mumbai       Mumbai        India      3   
14    C00014   Rangarappa  Bangalore    Bangalore        India      2   
15    C00025    Ravindran  Bangalore    Bangalore        India      2   
16    C00005     Sasikant     Mumbai       Mumbai        India      1   
17    C00006      Shilton    Torento      Torento       Canada      1   
18    C00017     Srinivas  Bangalore    Bangalore        India      2   
19    C00012       Steven   San Jose     San Jose          USA      1   
20    C00015       Stuart     London       London           UK      1   
21    C00011    Sundariya    Chennai      Chennai        India      3   
22    C00016   Venkatpati  Bangalore    Bangalore        India      2   
23    C00004      Winston    Brisban      Brisban    Australia      1   
24    C00019  Yearannaidu    Chennai      Chennai        India      1   

    OPENING_AMT  RECEIVE_AMT  PAYMENT_AMT  OUTSTANDING_AMT      PHONE_NO  \
0        5000.0       7000.0       6000.0           6000.0       BBBBSBB   
1        7000.0      11000.0       9000.0           9000.0  113-12345678   
2        5000.0       7000.0       9000.0           3000.0       DDNRDRH   
3        6000.0       4000.0       5000.0           5000.0       MMMMMMM   
4        4000.0       9000.0       7000.0           6000.0       FSDDSDF   
5        7000.0       7000.0       9000.0           5000.0       NHBGVFC   
6        6000.0       5000.0       7000.0           4000.0       BBBBBBB   
7        7000.0       7000.0       7000.0           7000.0       WERTGDF   
8        4000.0       6000.0       7000.0           3000.0       AAAABAA   
9        7000.0       7000.0       9000.0           5000.0       HJKORED   
10       8000.0       7000.0       7000.0           8000.0       MJYURFD   
11       3000.0       5000.0       2000.0           6000.0       CCCCCCC   
12       7000.0      11000.0       9000.0           9000.0       GHRDWSD   
13       8000.0       7000.0       3000.0          12000.0      Phone No   
14       8000.0      11000.0       7000.0          12000.0       AAAATGF   
15       5000.0       7000.0       4000.0           8000.0       AVAVAVA   
16       7000.0      11000.0       7000.0          11000.0  147-25896312   
17      10000.0       7000.0       6000.0          11000.0       DDDDDDD   
18       8000.0       4000.0       3000.0           9000.0       AAAAAAB   
19       5000.0       7000.0       9000.0           3000.0       KRFYGJK   
20       6000.0       8000.0       3000.0          11000.0       GFSGERS   
21       7000.0      11000.0       7000.0          11000.0       PPHGRTS   
22       8000.0      11000.0       7000.0          12000.0       JRTVFDD   
23       5000.0       8000.0       7000.0           6000.0       AAAAAAA   
24       8000.0       7000.0       7000.0           8000.0       ZZZZBFV   

   AGENT_CODE  
0        A008  
1        A002  
2        A008  
3        A009  
4        A006  
5        A005  
6        A003  

## [D.2] Mengelompokkan Data (Grouping)

### Data pada Tabel Customer Sebelum Dikelompokkan

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_penjualan.db')

# Membuat Query SQL
result  = connection.cursor()
command = "SELECT * FROM customer;"

# Eksekusi Query SQL
result.execute(command)

# Menampilkan Hasil Dalam Bentuk DataFrame
results      = result.fetchall();connection.close();
cols         = [column[0] for column in result.description]
results_data = pd.DataFrame.from_records(results, columns = cols)

results_data

CUST_CODE    CUST_NAME  CUST_CITY WORKING_AREA CUST_COUNTRY  GRADE  \
0     C00001      Micheal   New York     New York          USA      2   
1     C00002         Bolt   New York     New York          USA      3   
2     C00003       Martin    Torento      Torento       Canada      2   
3     C00004      Winston    Brisban      Brisban    Australia      1   
4     C00005     Sasikant     Mumbai       Mumbai        India      1   
5     C00006      Shilton    Torento      Torento       Canada      1   
6     C00007   Ramanathan    Chennai      Chennai        India      1   
7     C00008     Karolina    Torento      Torento       Canada      1   
8     C00009       Ramesh     Mumbai       Mumbai        India      3   
9     C00010      Charles  Hampshair    Hampshair           UK      3   
10    C00011    Sundariya    Chennai      Chennai        India      3   
11    C00012       Steven   San Jose     San Jose          USA      1   
12    C00013       Holmes     London       London           UK      2   
13    C00014   Rangarappa  Bangalore    Bangalore        India      2   
14    C00015       Stuart     London       London           UK      1   
15    C00016   Venkatpati  Bangalore    Bangalore        India      2   
16    C00017     Srinivas  Bangalore    Bangalore        India      2   
17    C00018      Fleming    Brisban      Brisban    Australia      2   
18    C00019  Yearannaidu    Chennai      Chennai        India      1   
19    C00020       Albert   New York     New York          USA      3   
20    C00021        Jacks    Brisban      Brisban    Australia      1   
21    C00022      Avinash     Mumbai       Mumbai        India      2   
22    C00023         Karl     London       London           UK      0   
23    C00024         Cook     London       London           UK      2   
24    C00025    Ravindran  Bangalore    Bangalore        India      2   

    OPENING_AMT  RECEIVE_AMT  PAYMENT_AMT  OUTSTANDING_AMT      PHONE_NO  \
0        3000.0       5000.0       2000.0           6000.0       CCCCCCC   
1        5000.0       7000.0       9000.0           3000.0       DDNRDRH   
2        8000.0       7000.0       7000.0           8000.0       MJYURFD   
3        5000.0       8000.0       7000.0           6000.0       AAAAAAA   
4        7000.0      11000.0       7000.0          11000.0  147-25896312   
5       10000.0       7000.0       6000.0          11000.0       DDDDDDD   
6        7000.0      11000.0       9000.0           9000.0       GHRDWSD   
7        7000.0       7000.0       9000.0           5000.0       HJKORED   
8        8000.0       7000.0       3000.0          12000.0      Phone No   
9        6000.0       4000.0       5000.0           5000.0       MMMMMMM   
10       7000.0      11000.0       7000.0          11000.0       PPHGRTS   
11       5000.0       7000.0       9000.0           3000.0       KRFYGJK   
12       6000.0       5000.0       7000.0           4000.0       BBBBBBB   
13       8000.0      11000.0       7000.0          12000.0       AAAATGF   
14       6000.0       8000.0       3000.0          11000.0       GFSGERS   
15       8000.0      11000.0       7000.0          12000.0       JRTVFDD   
16       8000.0       4000.0       3000.0           9000.0       AAAAAAB   
17       7000.0       7000.0       9000.0           5000.0       NHBGVFC   
18       8000.0       7000.0       7000.0           8000.0       ZZZZBFV   
19       5000.0       7000.0       6000.0           6000.0       BBBBSBB   
20       7000.0       7000.0       7000.0           7000.0       WERTGDF   
21       7000.0      11000.0       9000.0           9000.0  113-12345678   
22       4000.0       6000.0       7000.0           3000.0       AAAABAA   
23       4000.0       9000.0       7000.0           6000.0       FSDDSDF   
24       5000.0       7000.0       4000.0           8000.0       AVAVAVA   

   AGENT_CODE  
0        A008  
1        A008  
2        A004  
3        A005  
4        A002  
5        A004  
6        A010  

### Data pada Tabel Customer Setelah Dikelompokkan
Dikelompokkan berdasarkan Negara, dicari Total Piutang Pelanggan untuk setiap Negara

In [ ]:
# Buat Koneksi ke Database
connection = sqlite3.connect('data_penjualan.db')

# Membuat Query SQL
result  = connection.cursor()
command = """
             SELECT CUST_COUNTRY, SUM(OUTSTANDING_AMT) 
             FROM customer 
             GROUP BY CUST_COUNTRY;
          """

# Eksekusi Query SQL
result.execute(command)

# Menampilkan Hasil Dalam Bentuk DataFrame
results      = result.fetchall();connection.close();
cols         = [column[0] for column in result.description]
results_data = pd.DataFrame.from_records(results, columns = cols)

results_data

CUST_COUNTRY  SUM(OUTSTANDING_AMT)
0    Australia               18000.0
1       Canada               24000.0
2        India              101000.0
3           UK               29000.0
4          USA               18000.0

# [E] Latihan

Jawablah pertanyaan-pertanyaan berikut dengan SQL, menggunakan database data_penjualan.db
1. Tampilkan data customer yg mempunyai opening amount diatas 7000.

2. Tampilkan informasi order yg dilakukan melalui agensi dengan code A003.

3. Tampilkan data customer yg tinggal di New York, Bangalore, Mumbai, dan Torento.

4. Tampilkan informasi customer yang mempunyai payment amount diatas 7000 dan urutkan berdasarkan opening amountnya dari yg terbesar

5. Tampilkan data agensi yang tidak bekerja di london dan memiliki comission lebih dari 0.12
